go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,721 @@
+		if err != nil {
+			klog.Errorf("Failed to get address from link %s: %v", link.Attrs().Name, err)
+			continue
+		}
+		prefixes = append(prefixes, linkPrefixes...)
+		// create lookup table for later retrieval
+		for _, prefixFound := range linkPrefixes {
+			_, ipNet, err := net.ParseCIDR(prefixFound.String())
+			if err != nil {
+				klog.Errorf("Egress IP: skipping prefix %q due to parsing CIDR error: %v", prefixFound.String(), err)
+				continue
+			}
+			prefixLinks[ipNet.String()] = link
+		}
+	}
+	lpmTree := cidrtree.New(prefixes...)
+	addr, err := netip.ParseAddr(ip.String())
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to convert IP %s to netip addr: %v", ip.String(), err)
+	}
+	network, found := lpmTree.Lookup(addr)
+	if !found {
+		return false, nil, nil
+	}
+	if network.String() != expectedNetwork {
+		return false, nil, fmt.Errorf("EgressIP %s is was assigned to network %s but we got network %s. Mismatch therefore not assigning",
+			ip.String(), expectedNetwork, network.String())
+	}
+	return true, prefixLinks[network.String()], nil
+}
+
+func getNetworkInfoFromIP(ipStr string) (net.IP, *net.IPNet, bool, error) {
+	var err error
+	var ip net.IP
+	var ipNet *net.IPNet
+	var v6 bool
+	ip = net.ParseIP(ipStr)
+	if utilnet.IsIPv6(ip) {
+		v6 = true
+		ip, ipNet, err = net.ParseCIDR(fmt.Sprintf("%s/128", ipStr))
+	} else {
+		ip, ipNet, err = net.ParseCIDR(fmt.Sprintf("%s/32", ipStr))
+	}
+	return ip, ipNet, v6, err
+}
+
+func filterRouteByLinkTable(linkIndex, tableID int) (*netlink.Route, uint64) {
+	return &netlink.Route{
+			LinkIndex: linkIndex,
+			Table:     tableID,
+		},
+		netlink.RT_FILTER_OIF | netlink.RT_FILTER_TABLE
+}
+
+func filterRuleByPriority(priority int) (*netlink.Rule, uint64) {
+	return &netlink.Rule{
+			Priority: priority,
+		},
+		netlink.RT_FILTER_PRIORITY
+}
+
+func areNetlinkRulesEqual(r1, r2 netlink.Rule) bool {
+	if r1.Table != r2.Table {
+		return false
+	}
+	if r1.Priority != r2.Priority {
+		return false
+	}
+	if r1.Src != nil && r2.Src == nil {
+		return false
+	}
+	if r2.Src != nil && r1.Src == nil {
+		return false
+	}
+	if r1.Src != nil && r2.Src != nil && !r1.Src.IP.Equal(r2.Src.IP) {
+		return false
+	}
+	if r1.Src != nil && r2.Src != nil && r1.Src.Mask.String() != r2.Src.Mask.String() {
+		return false
+	}
+	if r1.Dst != nil && r2.Dst == nil {
+		return false
+	}
+	if r2.Dst != nil && r1.Dst == nil {
+		return false
+	}
+	if r1.Dst != nil && r2.Dst != nil && !r1.Dst.IP.Equal(r2.Dst.IP) {
+		return false
+	}
+	if r1.Dst != nil && r2.Dst != nil && r1.Dst.Mask.String() != r2.Dst.Mask.String() {
+		return false
+	}
+	if r1.Family != r2.Family {
+		return false
+	}
+	if r1.IifName != r2.IifName {
+		return false
+	}
+	if r1.OifName != r2.OifName {
+		return false
+	}
+	if r1.Mask != r2.Mask {
+		return false
+	}
+	if r1.Mark != r2.Mark {
+		return false
+	}
+	return true
+}
+
+func areEqual(s1, s2 []string) bool {
```

Comment:
Can you use `reflect.DeepEqual` instead?

Reply:
done